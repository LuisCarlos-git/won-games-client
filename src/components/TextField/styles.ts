import { TextFieldProps } from './index';
import styled, { css, DefaultTheme } from 'styled-components';

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.medium};
    color: ${theme.colors.black};
    cursor: pointer;
    display: block;
    margin-bottom: ${theme.spacings.xxsmall};
  `}
`;

const inputWrappermodifiers = {
  left: () => css`
    ${Input} {
      order: 2;
    }

    ${Icon} {
      order: 1;
    }
  `,
  right: () => css`
    ${Input} {
      order: 1;
      padding-left: 0;
    }

    ${Icon} {
      order: 2;
    }
  `
};

type InputWrapperProps = {
  error: boolean;
} & Pick<TextFieldProps, 'iconPosition'>;

export const InputWrapper = styled.div<InputWrapperProps>`
  ${({ theme, iconPosition, error }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${error ? theme.colors.error : theme.colors.lightGray};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    ${!!iconPosition && inputWrappermodifiers[iconPosition]}
  `}
`;

export const Icon = styled.div`
  ${({ theme }) => css`
    display: flex;
    width: 2.2rem;
    color: ${theme.colors.gray};
    & > svg {
      width: 100%;
    }
  `}
`;
const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};
      &::placeholder {
        color: currentColor;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${Label}, ${Icon} > svg {
      color: ${theme.colors.error};
    }
  `
};

export const Error = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    margin-top: ${theme.spacings.xxsmall};
    font-size: ${theme.font.sizes.small};
  `}
`;

export const Wrapper = styled.div<Pick<TextFieldProps, 'disabled' | 'error'>>`
  ${({ theme, disabled, error }) => css`
    ${!!disabled && wrapperModifiers.disabled(theme)}
    ${!!error && wrapperModifiers.error(theme)};
  `}
`;
