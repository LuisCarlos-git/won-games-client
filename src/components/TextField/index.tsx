import React, { InputHTMLAttributes, useState } from 'react';
import * as S from './styles';

export type TextFieldProps = {
  label?: string;
  labelFor?: string;
  initialValue?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
  disabled?: boolean;
  error?: string;
  onInput?: (value: string) => void;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  label,
  labelFor = '',
  initialValue = '',
  onInput,
  icon,
  iconPosition = 'left',
  disabled = false,
  error = '',
  ...rest
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;
    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  return (
    <S.Wrapper disabled={disabled} error={error}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper error={!!error} iconPosition={iconPosition}>
        {!!icon && <S.Icon>{icon}</S.Icon>}
        <S.Input
          disabled={disabled}
          id={labelFor}
          {...rest}
          onChange={onChange}
          value={value}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
};
export default TextField;
