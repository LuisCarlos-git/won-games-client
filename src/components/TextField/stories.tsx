import { Story, Meta } from '@storybook/react/types-6-0';
import TextField, { TextFieldProps } from '.';

export default {
  title: 'TextField',
  component: TextField
} as Meta;

export const Default: Story<TextFieldProps> = (args) => <TextField {...args} />;

Default.args = {
  label: 'E-mail',
  placeholder: 'luis@teste.com.br'
};
