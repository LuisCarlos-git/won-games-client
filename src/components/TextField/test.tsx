import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/tests/helper';
import { Email } from 'styled-icons/material-outlined';

import TextField from '.';

describe('<TextField />', () => {
  it('should render the label', () => {
    renderWithTheme(<TextField label="label" />);

    expect(screen.getByText(/label/i)).toBeInTheDocument();
  });

  it('should render the whithout label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByText(/label/i)).not.toBeInTheDocument();
  });

  it('should render the labelFor', () => {
    renderWithTheme(<TextField label="label" labelFor="labelFor" />);

    expect(screen.getByText(/label/i)).toHaveAttribute('for', 'labelFor');
  });

  it('should render the placeholder', () => {
    renderWithTheme(
      <TextField label="label" labelFor="labelFor" placeholder="any" />
    );

    expect(screen.getByLabelText(/label/i)).toHaveAttribute(
      'placeholder',
      'any'
    );
  });

  it('Changes values when typing', async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField onInput={onInput} label="label" labelFor="labelFor" />
    );

    const input = screen.getByRole('textbox');
    const text = 'this is my new text';

    userEvent.type(input, text);
    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toHaveBeenCalledWith(text);
  });

  it('Is accessible by tab', () => {
    renderWithTheme(
      <TextField label="TextField" labelFor="TextField" id="TextField" />
    );

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });

  it('icon version', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
      />
    );

    expect(screen.getByLabelText(/textfield/i)).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('icon default position (left)', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
      />
    );

    expect(screen.getByTestId(/icon/i).parentElement).toHaveStyle({
      order: '1'
    });
  });

  it('icon position (right)', () => {
    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId(/icon/i).parentElement).toHaveStyle({
      order: '2'
    });
  });

  it('input disabled', async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
        disabled
      />
    );

    const input = screen.getByRole('textbox');
    expect(input).toBeDisabled();

    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it('Check the render of erros', async () => {
    const { container } = renderWithTheme(
      <TextField
        label="TextField"
        labelFor="TextField"
        id="TextField"
        icon={<Email data-testid="icon" />}
        error="Digite um e-mail valido!"
      />
    );

    expect(screen.getByText(/Digite um e-mail valido!/i)).toBeInTheDocument();

    expect(screen.getByTestId('icon')).toHaveStyle({
      color: '#e31963'
    });
    expect(screen.getByText(/textfield/i)).toHaveStyle({
      color: '#e31963'
    });

    expect(screen.getByRole('textbox').parentElement).toHaveStyle({
      'border-color': '#e31963'
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
