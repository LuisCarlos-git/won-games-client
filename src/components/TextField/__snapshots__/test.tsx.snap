// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Check the render of erros 1`] = `
.c6 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.6rem;
  color: #030517;
  cursor: pointer;
  display: block;
  margin-bottom: 0.8rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #e31963;
}

.c3:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c3 .c7 {
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

.c3 .c4 {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.2rem;
  color: #8F8F8F;
}

.c5 > svg {
  width: 100%;
}

.c9 {
  color: #e31963;
  margin-top: 0.8rem;
  font-size: 1.4rem;
}

.c0 .c1,
.c0 .c4 > svg {
  color: #e31963;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3"
  >
    <div
      class="c4 c5"
    >
      <svg
        aria-hidden="true"
        class="c6"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0l-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
        />
      </svg>
    </div>
    <input
      class="c7 c8"
      id="TextField"
      value=""
    />
  </div>
  <div
    class="c9"
  >
    Digite um e-mail valido!
  </div>
</div>
`;
